src/hugr/validate.rs

```diff
@@ -173,6 +269,19 @@ impl<'a> ValidationContext<'a> {
             });
         }
 
+        // Ignore subport indices assuming that we only care about dataflow nodes
+        for (port_index, link_index) in links.clone().into_iter().map(|(a, b)| (a.port(), b.port()))
```

Comment:
Why not move this to the `for (subport, link) in links {` loop below?
Then you know that the edge goes from `subport` to `link`, so you don't need the extra logic handling directions in `check_resources_compatible`.

Reply:
Good spot!
Are you saying that `graph.port_links(x)` always returns pairs in order (i.e. `(source, target)`)? Even if `x` is the target?